Race condition: Bus should lock subscriptions array while reading in Send method

I have ran into a problem with the `Bus.Send()` method in `stdbus/` implementation. 
It raised error when tried read `.key` from one of the subscriptions which was nil.

This bug may happen when run concurrently: 
a) remove `subcription` from `array of subscriptions` in bus and 
b) loop over this  `array of subscriptions` in Send() method. 

For check this problem, you can run single test with `-race` flag for execute with Go's race detector:
```
cd stdbus 
go test -race -run TestEvents
```
Output will be:
```
==================
WARNING: DATA RACE
Read at 0x00c00006f350 by goroutine 11:
  github.com/devishot/ari/stdbus.(*bus).Send()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus.go:50 +0x59
  github.com/devishot/ari/stdbus.TestEventsDataRace.func2()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus_test.go:208 +0xaa

Previous write at 0x00c00006f350 by goroutine 10:
  github.com/devishot/ari/stdbus.(*bus).add()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus.go:82 +0xe8
  github.com/devishot/ari/stdbus.(*bus).Subscribe()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus.go:75 +0x186
  github.com/devishot/ari/stdbus.TestEventsDataRace.func1()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus_test.go:194 +0x21a

Goroutine 11 (running) created at:
  github.com/devishot/ari/stdbus.TestEventsDataRace()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus_test.go:204 +0x377
  testing.tRunner()
      C:/Go/src/testing/testing.go:827 +0x169

Goroutine 10 (running) created at:
  github.com/devishot/ari/stdbus.TestEventsDataRace()
      C:/Users/stamkulov/Documents/Kolesa/go/src/github.com/devishot/ari/stdbus/bus_test.go:187 +0x33f
  testing.tRunner()
      C:/Go/src/testing/testing.go:827 +0x169
==================
--- FAIL: TestEventsDataRace (0.00s)
    testing.go:771: race detected during execution of test
FAIL
exit status 1
FAIL    github.com/devishot/ari/stdbus  0.426s
```


- [ ]  Please have a look my PR for this problem. It uses mutex Lock for `array of subscriptions` in `Bus.Send()`

--- a/stdbus/bus.go
+++ b/stdbus/bus.go
@@ -89,7 +89,6 @@ func (b *bus) add(s *subscription) {
 // remove deletes the given subscription from the bus
 func (b *bus) remove(s *subscription) {
 	b.rwMux.Lock()
-	defer b.rwMux.Unlock()
 	for i, si := range b.subs {
 		if s == si {
 			// Subs are pointers, so we have to explicitly remove them
@@ -97,9 +96,10 @@ func (b *bus) remove(s *subscription) {
 			b.subs[i] = b.subs[len(b.subs)-1] // replace the current with the end
 			b.subs[len(b.subs)-1] = nil       // remove the end
 			b.subs = b.subs[:len(b.subs)-1]   // lop off the end
-			return
+			break
 		}
 	}
+	b.rwMux.Unlock()
 }
 
 // A Subscription is a wrapped channel for receiving

--- a/stdbus/bus.go
+++ b/stdbus/bus.go
@@ -16,7 +16,7 @@ var subscriptionEventBufferSize = 100
 type bus struct {
 	subs []*subscription // The list of subscriptions
 
-	mu sync.Mutex
+	rwMux sync.RWMutex
 
 	closed bool
 }
@@ -45,7 +45,7 @@ func (b *bus) Close() {
 // Send sends the message to the bus
 func (b *bus) Send(e ari.Event) {
 	var matched bool
-	b.mu.Lock()
+	b.rwMux.RLock()
 
 	// Disseminate the message to the subscribers
 	for _, s := range b.subs {
@@ -68,7 +68,7 @@ func (b *bus) Send(e ari.Event) {
 		}
 	}
 
-	b.mu.Unlock()
+	b.rwMux.RUnlock()
 }
 
 // Subscribe returns a subscription to the given list
@@ -81,15 +81,15 @@ func (b *bus) Subscribe(key *ari.Key, eTypes ...string) ari.Subscription {
 
 // add appends a new subscription to the bus
 func (b *bus) add(s *subscription) {
-	b.mu.Lock()
+	b.rwMux.Lock()
 	b.subs = append(b.subs, s)
-	b.mu.Unlock()
+	b.rwMux.Unlock()
 }
 
 // remove deletes the given subscription from the bus
 func (b *bus) remove(s *subscription) {
-	b.mu.Lock()
-	defer b.mu.Unlock()
+	b.rwMux.Lock()
+	defer b.rwMux.Unlock()
 	for i, si := range b.subs {
 		if s == si {
 			// Subs are pointers, so we have to explicitly remove them

--- a/stdbus/bus.go
+++ b/stdbus/bus.go
@@ -45,6 +45,7 @@ func (b *bus) Close() {
 // Send sends the message to the bus
 func (b *bus) Send(e ari.Event) {
 	var matched bool
+	b.mu.Lock()
 
 	// Disseminate the message to the subscribers
 	for _, s := range b.subs {
@@ -66,6 +67,8 @@ func (b *bus) Send(e ari.Event) {
 			}
 		}
 	}
+
+	b.mu.Unlock()
 }
 
 // Subscribe returns a subscription to the given list

